### ISSUE TITLE
document what lints have quick-fixes

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#2650]
See also: https://github.com/dart-lang/lints/issues/12.

In the generated docs, it would be great to document if a given lint has an associated quick-fix.  (And maybe bulk-fix?)

@parlough, @kwalrath: ideas for how to present this information most welcome!

**Comment by bwilkerson** [comment]
I made this comment on the other thread, but I'll repeat it here because it's a different audience.

I've wondered a few time whether we might want to add something to the documentation for the lints (and other diagnostics) to indicate that there are fixes available for them. Perhaps a generated line at the end for the form

Fixable using dart fix or in the IDE using '<fix label>'.

with support for a list if there's more than one fix.

**Comment by bernaferrari** [comment]
I'm having the same problem you just described @bwilkerson. I added flutter_lints and now half my codebase begs for const in widgets. However, dart fix doesn't handle these (even though the IDE has the yellow lamp). I also expected dart fix to handle unused imports, but it doesn't. I believe you could make dart fix more powerful, even if it required an `--aggressive` command. I prefer to fix 200 issues even if it creates 5 than having to manually fix 200 things (and potentially creating more than 5).

**Comment by bwilkerson** [comment]
There is definitely room to improve the tool.

> ... begs for const in widgets. However, dart fix doesn't handle these (even though the IDE has the yellow lamp).

The lints would probably be because of the `prefer_const_constructors` rule. The reason that particular fix isn't enabled is because the fix isn't clever enough to notice when a `const` doesn't need to be added in one place because an outer constructor invocation had a `const` added to it, so it adds unnecessary code (which turns into lints if you also have the `unnecessary_const` rule enabled). We debated about whether it would be better to have fewer fixes enabled or to require users to run `dart fix` multiple times, and we decided to enable fewer fixes. Sounds like you might have preferred the other option.

Of course, the better solution (which we didn't have time to implement earlier) would be to fix the fix generator so that the problem doesn't arise. That's definitely something we're hoping to get back to at some point.

> I also expected dart fix to handle unused imports, but it doesn't.

That's because of another conflict. We recently added the ability to upgrade uses of some Flutter APIs when the APIs had changed. For example, if a class is renamed we can update the old name to the new name even if the old class has been deleted (as opposed to just being marked as `@deprecated`). In such a scenario the import might appear to be unused. Unfortunately, at that point the two fixes don't play nicely together. The fix to update the name replaces the name and sees that the import is still there, then the unused import fix removes the import (not realizing that it's no longer unused) and the result is that a now necessary import is removed and the new name is unknown.

We have a plan for how to fix this as well. The plan is to apply all the other fixes first, then re-analyze the code and only fix the unused imports after all other changes have been applied. I'm not sure when we'll be able to work on that, but we decided that until we could fix the behavior it was better to leave unused imports unfixed.

**Comment by bernaferrari** [comment]
> const doesn't need to be added in one place because an outer constructor invocation had a const added to it, so it adds unnecessary code (which turns into lints if you also have the unnecessary_const rule enabled)

Yeah, I saw that when fixing my issues. I always think, if the tool has some issues, it could be called again and again until it fixes all (or, like, run 3 times to guarantee it fixed everything. I can wait).

Thanks for explaining how unused imported works, I had no idea it was so complex!

**Comment by parlough** [comment]
I like your solution for identifying the fixes, definitely easier, thanks. If we added a mention to the docs similar to this, is it relevant to list the labels of fixes? I can't think of an example, but I believe a lint could potentially have multiple fixes.

**Comment by bwilkerson** [comment]
I would go with stable for now. If we wanted we could expand that later to include information about which channel users need to be on to be able to use the fix, but I'm not sure how useful that is.

There's interest earlier in this thread for separately indicating whether the fix is available via `dart fix`, but that would be even harder to provide (because it would require parsing the implementations of all of the fixes associated with the lints).

There's another approach that just occurred to me. We could write a tool in the analysis_server package that generated a file containing the information we're currently parsing files to get. The linter could then read that file rather than parsing any files in server. The advantage is that the generator tool could load the actual data rather than depending on parsing an assumed structure, which would make it less brittle and would allow bulk fix (and maybe fix-in-file) information to be easier to get.

**Comment by pq** [comment]
Motivated by a discussion w/ @jcollins-g, @keertip, @srawlins and @bwilkerson today, I'm going to look at wrapping this up.  My current thinking for an initial (easy) deliverable would be to:

- [ ] update [`machine.json`](https://dart-lang.github.io/linter/lints/machine/rules.json) with fix status ("has fix", "no fix", "under evaluation") as pulled from [`error_fix_status`](https://github.com/dart-lang/sdk/blob/main/pkg/analysis_server/lib/src/services/correction/error_fix_status.yaml)
- [ ] add a badge with fix status to generated lint docs (like the ones for rule sets)

Feedback here would be most welcome -- in particular from @sfshaza2 and @parlough who might have thoughts about how this will play with plans for consolidating lint rule documentation in the main Dart SDK dev docs down the road.

**Comment by parlough** [comment]
That sounds good to me and we could definitely surface that information on the dart.dev version of the docs.

I guess the question is around @bwilkerson's plan to consolidate linter documentation with the other diagnostics, as I don't believe we document their fix status yet either.

**Comment by bwilkerson** [comment]
Right. I've started a slow-moving background task to re-write the lint docs in the style of the Diagnostic Messages page, and @sfshaza2 and I have talked a couple of times about wanting to add some of the lint badges to the Diagnostic Messages page, but we don't yet have a concrete proposal for which badges to include or what the actual page content would look like.

@pq and I have also talked a couple of times about the best way to implement this: whether to leave the docs where they are today, whether to move them into a YAML doc like the non-lint docs, or something else. That hasn't been decided yet either. One open question is whether we want to preserve the look of the current lint docs in the current doc pages or whether we want to update them in both places at the same time.
